# Text Similarity and Clustering

Previous weeks have covered several techniques of analyzing text and extracting interesting insights. We have looked at supervised machine learning (ML) techniques that are used to classify or categorize text documents into several pre-assumed categories.

Today, we will be looking at several other techniques and use-cases that leverage unsupervised learning and information retrieval concepts, mainly, understanding text similarity and clustering. We will discuss some important concepts related to information retrieval, document similarity measures, and machine learning.

One constraint in text classification is that we need some training data with manually labeled categories because we use supervised learning algorithms to build our classification model. The efforts of building this dataset are definitely not easy, because to build a good model, you need a sizeable amount of training data. For this, we need to spend time and manual effort in labeling data, building a model, and then finally using it to classify new documents. Can we instead make the machine do it? Yes, as a matter of fact, we can. This chapter specifically addresses looking at the content of text documents, analyzing their similarity using various measures, and clustering similar documents together.

Text data is unstructured and highly noisy. We get the benefits of well-labeled training data and supervised learning when performing text classification. But document clustering is an unsupervised learning process, where we are trying to segment and categorize documents into separate categories by making the machine learn about the various text documents, their features, similarities, and the differences among them. This makes document clustering more challenging, albeit interesting.

We will focus on several concepts related to text similarity, distance metrics, and unsupervised ML algorithms to answer the following questions
<ul>
<li>How do we measure similarity between documents?</li>
<li>How can we use distance measures to find the most relevant documents?</li>
<li>When is a distance measure called a metric?</li>
<li>How do we cluster or group similar documents?</li>
<li>Can we visualize document clusters?</li>
</ul>

## Information Retrieval (IR)

Information retrieval (IR) is the process of retrieving or fetching relevant sources of information from a corpus or set of entities that hold information based on some demand. For example, it could be a query or search that users enter in a search engine and then get relevant search items pertaining to their query. In fact, search engines are the most popular use-case or application of IR.

The relevancy of documents with information compared to the demand can be measured in several ways. It can include looking for specific keywords from the search text or using some similarity measures to see the similarity rank or score of the documents with respect to the entered query. This makes is quite different from string matching or matching regular expressions because more than often the words in a search string can have different order, context, and semantics in the collection of documents (entities), and these words can even have multiple different resolutions or possibilities based on synonyms, antonyms, and negation modifiers.

## Feature Engineering

Feature engineering or feature extraction is something which you know quite well by now. Methods like Bag of Words, TF-IDF, and word vectorization models are typically used to represent or model documents in the form of numeric vectors so that applying mathematical or machine learning techniques become much easier. You can use various document representations using these feature-extraction techniques or even map each letter or a word to a corresponding unique numeric identifier.

## Similarity Measures

Similarity measures are used frequently in text similarity analysis and clustering. Any similarity or distance measure usually measures the degree of closeness between two entities, which can be any text format like documents, sentences, or even terms. This measure of similarity can be useful in identifying similar entities and distinguishing clearly different entities from each other. Similarity measures are very effective, and sometimes choosing the right measure can make a lot of difference in the performance of your final analytics system. Various scoring or ranking algorithms have also been invented based on these distance measures.

There are several distance measures that measure similarity, and we will be covering several of them. However, an important thing to remember is that all distance measures of similarity are not distance metrics of similarity.

Consider a distance measure <code>d</code> and two entities (say they are documents in our context) <code>x</code> and <code>y</code>. The distance between x and y, which is used to determine the degree of similarity between them, can be represented as <code>d(x, y)</code>, but the measure d can be called as a distance metric of similarity if and only if it satisfies the following four conditions:

<ol>
<li>The distance measured between any two entities, say x and y, must be always non-negative, that is, d(x, y) $\geq$ 0</li>
<li>The distance between two entities should always be zero if and only if they are both identical, that is, d(x, y) $\geq$ 0 <i>iff</i> x=y</li>
<li>This distance measure should always be symmetric, which means that the distance from x to y is always the same as the
distance from <code>y</code> to <code>x</code>. Mathematically this is represented as <code>d(x, y) = d(y, x)</code></li>
<li>This distance measure should satisfy the triangle inequality property, which can be mathematically represented
    d(x, z) $\leq$ d(x, y) + d(y, z)
    </li>
</ol>

## Unsupervised Machine Learning Algorithms

<code>Unsupervised machine learning algorithms</code> are the family of ML algorithms that try to discover latent hidden structures and patterns in data from their various attributes and features. Besides this, several unsupervised learning algorithms are also used to reduce the feature space, which is often of a higher dimension to one with a lower dimension. The data on which these algorithms operate is essentially unlabeled data that does not have any pre-determined category or class. We apply these algorithms with the intent of finding patterns and distinguishing features that might help us in grouping various data points into groups or clusters. These algorithms are popularly known as <code>clustering algorithms</code>.

### Text Normalization

### Feature Extraction

In [1]:
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
def build_feature_matrix(documents, feature_type='frequency', ngram_range=(1, 1), min_df=0.0, max_df=1.0):
    feature_type = feature_type.lower().strip()

    if feature_type == 'binary':
        vectorizer = CountVectorizer(binary=True, min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    elif feature_type == 'frequency':
        vectorizer = CountVectorizer(binary=False, min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    elif feature_type == 'tfidf':
        vectorizer = TfidfVectorizer(min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    else:
        raise Exception("Wrong feature type entered. Possible values:'binary', 'frequency','tfidf'")

    feature_matrix = vectorizer.fit_transform(documents).astype(float)
    return vectorizer, feature_matrix

You can see from the function definition that we have capabilities for Bag of Words frequency, occurrences, and also TF-IDF–based features. The new additions in this function include the addition of the min_df, max_df, and ngram_range parameters and also accepting them as optional arguments. N-grams are contiguous sequences of "n" items, typically words.

<ul>
<li>ngram_range is useful when we want to add bigrams, trigrams, and so on as additional features.</li>
<li>min_df parameter can be expressed by a threshold value within a range of [0.0, 1.0] and it will ignore terms as features that will have a document frequency strictly lower than the input threshold value.</li>    
<li>max_df parameter can also be expressed by a threshold value within a range of [0.0, 1.0] and it will ignore terms as features that will have a document frequency strictly higher than the input threshold value. </li>
</ul>

<img src="https://github.com/irungus/TUDA/blob/main/img/grams.png?raw=1" width="350" height="200">

The intuition behind this would be that these words, if they occur in almost all the documents, tend to have little value that would help us in distinguishing among various types of documents.

We will now deep dive into the various techniques for text similarity
    
    


## Text Similarity

The main objective of text similarity is to analyze and measure how two entities of text are close or far apart from each other. These entities of text can be simple tokens or terms, like words, or whole documents, which may include sentences or paragraphs of text. There are various ways of analyzing text similarity, and we can classify the intent of text similarity broadly into the following two areas:
<ul>
<li><code>Lexical similarity</code>: This involves observing the contents of the <code>text documents with regard to syntax, structure, and <b>content</b></code> and measuring their similarity based on these parameters</li>    
<li><code>Semantic similarity</code>: This involves trying to find out the <code>semantics, meaning, and <b>context</b></code> of the documents and then trying to see how close they are to each other. Dependency grammars and entity recognition are handy tools that can help in this.</li>
</ul>



Note that the most popular area is lexical similarity, because the techniques are more straightforward, easy to implement, and you can also cover several parts of semantic similarity using simple models like the Bag of Words. Usually distance metrics will be used to measure similarity scores between text entities, and we will be mainly covering the following two broad areas of text similarity:
<ul>
<li><code>Term similarity</code>: Here we will measure similarity between individual tokens or words.</li>
<li><code>Document similarity</code>: Here we will be measuring similarity between entire text documents.</li>
</ul>

The idea is to implement and use several distance metrics and see how we can measure and analyze similarity among entities that are just simple words, and then how things change when we measure similarity among documents that are groups of individual words.

### Term Similarity

<code>Term similarity</code>: Measures similarity between individual tokens or words.

Several applications and use-cases like autocompleters, spell check, and correctors use some of these techniques to correct misspelled terms.

Here we will be taking a couple of words and measuring the similarity between then using different word representations as
well as distance metrics. The word representations we will be using are as follows:

<ul>
<li>Character vectorization</li>
<li>Bag of Characters vectorization</li>
</ul>

####  Character vectorization

Character vectorization is an extremely simple process of just mapping each character of the term to a corresponding unique number. We can do that using the function depicted in the following snippet: The function takes input a list of words or terms and returns the corresponding character vectors for the words.

In [2]:
import numpy as np
def vectorize_terms(terms):
    terms = [term.lower() for term in terms]
    terms = [np.array(list(term)) for term in terms]
    terms = [np.array([ord(char) for char in term]) for term in terms]
    return terms

In [3]:
vectorize_terms('pet')

[array([112]), array([101]), array([116])]

In [4]:
vectorize_terms('pot')

[array([112]), array([111]), array([116])]

In [5]:
vectorize_terms('potato')

[array([112]),
 array([111]),
 array([116]),
 array([97]),
 array([116]),
 array([111])]

In [6]:
root = 'Believe'
term1 = 'beleive'
term2 = 'bargain'
term3 = 'Elephan'

In [8]:
terms = [root, term1, term2, term3]

In [9]:
# Character vectorization
vec_root, vec_term1, vec_term2, vec_term3 = vectorize_terms(terms)

In [ ]:
print ('root: {} \nterm1: {} \nterm2: {} \nterm3: {}'.format(vec_root, vec_term1, vec_term2,
                                                             vec_term3))

root: [ 98 101 108 105 101 118 101] 
term1: [ 98 101 108 101 105 118 101] 
term2: [ 98  97 114 103  97 105 110] 
term3: [101 108 101 112 104  97 110]


#### Bag of Characters

Bag of Characters vectorization is very similar to the Bag of Words model except here we compute the frequency of each character in the word. Sequence or word orders are not taken into account. The following function helps in computing this.

In that function, we take in a list of words or terms and then extract the unique characters from all the words. This becomes our feature list, just like we do in Bag of Words, where instead of characters, unique words are our features. Once we have this list of unique_chars, we get the count for each of the characters in each word and build our Bag of Characters vectors

In [10]:
#from scipy.stats import itemfreq
import numpy as np
def boc_term_vectors(word_list):
    word_list = [word.lower() for word in word_list]
    unique_chars = np.unique(np.hstack([list(word) for word in word_list]))
    word_list_term_counts = [{char: count for char, count in zip(*np.unique(list(word), return_counts=True))} for word in word_list]
    boc_vectors = [np.array([int(word_term_counts.get(char, 0)) for char in unique_chars]) for word_term_counts in word_list_term_counts]
    return list(unique_chars), boc_vectors

In [ ]:
boc_term_vectors('pet pot cat')

([np.str_(' '),
  np.str_('a'),
  np.str_('c'),
  np.str_('e'),
  np.str_('o'),
  np.str_('p'),
  np.str_('t')],
 [array([0, 0, 0, 0, 0, 1, 0]),
  array([0, 0, 0, 1, 0, 0, 0]),
  array([0, 0, 0, 0, 0, 0, 1]),
  array([1, 0, 0, 0, 0, 0, 0]),
  array([0, 0, 0, 0, 0, 1, 0]),
  array([0, 0, 0, 0, 1, 0, 0]),
  array([0, 0, 0, 0, 0, 0, 1]),
  array([1, 0, 0, 0, 0, 0, 0]),
  array([0, 0, 1, 0, 0, 0, 0]),
  array([0, 1, 0, 0, 0, 0, 0]),
  array([0, 0, 0, 0, 0, 0, 1])])

In [11]:
root = 'Believe'
term1 = 'beleive'
term2 = 'bargain'
term3 = 'Elephan'

In [12]:
terms = [root, term1, term2, term3]

In [13]:
# Bag of characters vectorization
features, (boc_root, boc_term1, boc_term2, boc_term3) = boc_term_vectors(terms)

In [14]:
print ('Features:', features)
print ('root: {} \nterm1: {} \nterm2: {} \nterm3: {}'.format(boc_root, boc_term1, boc_term2, boc_term3))

Features: ['a', 'b', 'e', 'g', 'h', 'i', 'l', 'n', 'p', 'r', 'v']
root: [0 1 3 0 0 1 1 0 0 0 1] 
term1: [0 1 3 0 0 1 1 0 0 0 1] 
term2: [2 1 0 1 0 1 0 1 0 1 0] 
term3: [1 0 2 0 1 0 1 1 1 0 0]


### Distance Metrics to Compute Similarity

Thus you can see how we can easily transform text terms into numeric vector representations. We will now be using several distance metrics to compute similarity between the root word and the other three words mentioned in the preceding snippet. There are a lot of distance metrics out there that you can use to compute and measure similarities. We will be covering the following:
<ul>
<li>Hamming distance</li>
<li>Manhattan distance</li>
<li>Euclidean distance</li>
<li>Levenshtein edit distance</li>
<li>Cosine distance and similarity</li>
</ul>

We will set up some necessary variables storing the root term, the other terms with which its similarity will be measures, and their various vector representations using the following snippet:

In [15]:
root_term = root
root_vector = vec_root
root_boc_vector = boc_root
terms = [term1, term2, term3]
vector_terms = [vec_term1, vec_term2, vec_term3]
boc_vector_terms = [boc_term1, boc_term2, boc_term3]

In [16]:
print(root_term)
print(root_vector)
print(root_boc_vector)
print(terms)
print(vector_terms)
print(boc_vector_terms)

Believe
[ 98 101 108 105 101 118 101]
[0 1 3 0 0 1 1 0 0 0 1]
['beleive', 'bargain', 'Elephan']
[array([ 98, 101, 108, 101, 105, 118, 101]), array([ 98,  97, 114, 103,  97, 105, 110]), array([101, 108, 101, 112, 104,  97, 110])]
[array([0, 1, 3, 0, 0, 1, 1, 0, 0, 0, 1]), array([2, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0]), array([1, 0, 2, 0, 1, 0, 1, 1, 1, 0, 0])]


We are now ready to start computing similarity metrics and will be using the preceding terms and their vector representations to measure similarities

#### Hamming Distance

The Hamming distance is a very popular distance metric used frequently in information theory and communication systems. It is distance measured between two strings under the assumption that they are of equal length. Formally, it is defined as the number of positions that have different characters or symbols between two strings of equal length. Considering two terms u and v of length n, we can mathematically denote Hamming distance as:

<img src="https://github.com/irungus/TUDA/blob/main/img/fm1.png?raw=1" width="200" height="150">

and you can also normalize it if you want by dividing the number of mismatches by the total length of the terms to give the normalized hamming distance, which is represented as:

whereas you already know n denotes the length of the terms.

<img src="https://github.com/irungus/TUDA/blob/main/img/fm2.png?raw=1" width="250" height="200">

The following function computes the Hamming distance between two terms and also has the capability to compute the normalized distance:

In [24]:
def hamming_distance(u, v, norm=False):
    if u.shape != v.shape:
        raise ValueError('The vectors must have equal lengths.')
    return (u != v).sum() if not norm else (u != v).mean()

We will now measure the Hamming distance between our root term and the other terms using the following code snippet:

In [25]:
# compute Hamming distance
for term, vector_term in zip(terms, vector_terms):
    print('Hamming distance between root: {} and term: {} is {}'
          .format(root_term,term,hamming_distance(root_vector, vector_term, norm=False)))

Hamming distance between root: Believe and term: beleive is 2
Hamming distance between root: Believe and term: bargain is 6
Hamming distance between root: Believe and term: Elephan is 7


In [26]:
vector_terms

[array([ 98, 101, 108, 101, 105, 118, 101]),
 array([ 98,  97, 114, 103,  97, 105, 110]),
 array([101, 108, 101, 112, 104,  97, 110])]

In [27]:
# compute normalized Hamming distance
for term, vector_term in zip(terms, vector_terms):
    print('Normalized Hamming distance between root: {} and term: {} is {}'
          .format(root_term,term,round(hamming_distance(root_vector, vector_term, norm=True), 2)))

Normalized Hamming distance between root: Believe and term: beleive is 0.29
Normalized Hamming distance between root: Believe and term: bargain is 0.86
Normalized Hamming distance between root: Believe and term: Elephan is 1.0


You can see from the preceding output that terms 'Believe' and 'believe' ignoring their case are most similar to each other with the Hamming distance of 2 or 0.29, compared to the term 'bargain' giving scores of 6 or 0.86 (here, the smaller the score, the more similar are the terms). The term 'Elephant' throws an exception because the length of that term (term3) is 8 compared to length 7 of the root term 'Believe', hence Hamming distance can’t be computed because the base assumption of strings being of equal length is violated.

#### Manhattan Distance

The Manhattan distance metric is similar to the Hamming distance conceptually, where instead of counting the number of mismatches, we subtract the difference between each pair of characters at each position of the two strings. Formally, Manhattan distance is also known as city block distance, L1 norm, taxicab metric and is defined as the distance between two points in a grid based on strictly horizontal or vertical paths instead of the diagonal distance conventionally calculated by the Euclidean distance metric. Mathematically it can be denoted as

<img src="https://github.com/irungus/TUDA/blob/main/img/fm3.png?raw=1" width="200" height="150">

where u and v are the two terms of length n. The same assumption of the two terms having equal length from Hamming distance holds good here. We can also compute the normalized Manhattan distance by dividing the sum of the absolute differences by the term length. This can be denoted by

<img src="https://github.com/irungus/TUDA/blob/main/img/fm4.png?raw=1" width="230" height="170">

where n is the length of each of the terms u and v. The following function helps us in implementing Manhattan distance with the capability to also compute the normalized Manhattan distance:

In [21]:
def manhattan_distance(u, v, norm=False):
    if u.shape != v.shape:
        raise ValueError('The vectors must have equal lengths.')
    return abs(u - v).sum() if not norm else abs(u - v).mean()

We will now compute the Manhattan distance between our root term and the other terms using the previous function, as shown in the following code snippet:

In [22]:
for term, vector_term in zip(terms, vector_terms):
    print('Manhattan distance between root: {} and term: {} is {}'
          .format(root_term,term, manhattan_distance(root_vector,vector_term, norm=False)))

Manhattan distance between root: Believe and term: beleive is 8
Manhattan distance between root: Believe and term: bargain is 38
Manhattan distance between root: Believe and term: Elephan is 57


In [23]:
# compute normalized Manhattan distance
for term, vector_term in zip(terms, vector_terms):
    print('Normalized Manhattan distance between root: {} and term: {} is {}'
          .format(root_term,term,round(manhattan_distance(root_vector, vector_term,norm=True),2)))

Normalized Manhattan distance between root: Believe and term: beleive is 1.14
Normalized Manhattan distance between root: Believe and term: bargain is 5.43
Normalized Manhattan distance between root: Believe and term: Elephan is 8.14


From those results you can see that as expected, the distance between 'Believe' and 'believe' ignoring their case is most similar to each other, with a score of 8 or 1.14, as compared to 'bargain', which gives a score of 38 or 5.43 (here the smaller the score, the more similar the words). The term 'Elephant' yields an error because it has a different length compared to the base term just as we noticed earlier when computing Hamming distances.

#### Euclidean Distance

We briefly mentioned the Euclidean distance when comparing it with the Manhattan distance in the earlier section. Formally, the Euclidean distance is also known as the Euclidean norm, L2 norm, or L2 distance and is defined as the shortest straight-line distance between two points. Mathematically this can be denoted as

<img src="https://github.com/irungus/TUDA/blob/main/img/fm5.png?raw=1" width="230" height="170">

where the two points u and v are vectorized text terms in our scenario, each having length n. The following function helps us in computing the Euclidean distance between two terms:

In [ ]:
def euclidean_distance(u, v):
    if u.shape != v.shape:
        raise ValueError('The vectors must have equal lengths.')
    distance = np.sqrt(np.sum(np.square(u - v)))
    return distance

We can now compare the Euclidean distance among our terms by using the preceding function as depicted in the following code snippet:

In [ ]:
for term, vector_term in zip(terms, vector_terms):
    print('Euclidean distance between root: {} and term: {} is {}'
          .format(root_term,term, round(euclidean_distance(root_vector, vector_term),2)))

Euclidean distance between root: Believe and term: beleive is 5.66
Euclidean distance between root: Believe and term: bargain is 17.94
Euclidean distance between root: Believe and term: Elephan is 26.21


From the preceding outputs you can see that the terms 'Believe' and 'believe' are the most similar with a score of 5.66 compared to 'bargain' giving us a score of 17.94, and 'Elephant' throws a ValueError because the base assumption that strings being compared should have equal lengths holds good for this distance metric also.

<b>Disclaimer:</b> So far, all the distance metrics we have used work on strings or terms of the same length and fail when they are not of equal length. So how do we deal with this problem? We will now look at a couple of distance metrics that work even with strings of unequal length to measure similarity.

#### Levenshtein Edit Distance

The Levenshtein edit distance, often known as just Levenshtein distance, belongs to the family of edit distance–based metrics and is used to measure the distance between two sequence of strings based on their differences—similar to the concept behind Hamming distance. The Levenshtein edit distance between two terms can be defined as the minimum number of edits needed in the form of additions, deletions, or substitutions to change or convert one term to the other. These substitutions are character-based substitutions, where a single character can be edited in a single operation. Also, as mentioned before, the length of the two terms need not be equal here. Mathematically, we can represent the Levenshtein edit distance between u and v are our two terms where |u| and |v| are their lengths, in the formulae below:

<img src="https://github.com/irungus/TUDA/blob/main/img/fm6.png?raw=1" width="450" height="400">

where i and j are basically indices for the terms u and v. The third equation in the minimum above has a cost function denoted by $C_{u_i \neq u_j}$ such that it has the following conditions

<img src="https://github.com/irungus/TUDA/blob/main/img/fm7.png?raw=1" width="230" height="170">

and this denotes the indicator function, which depicts the cost associated with twocharacters being matched for the two terms (the equation represents the match or mismatch operation)

The first equation in the previous minimum stands for the deletion operation, and the second equation represents the insertion operation. The function $ld_{u,v}(i, j)$ thus covers all the three operations of insertion, deletion, and addition as we mentioned earlier and it denotes the Levenshtein distance as measured between the first i characters for the term u and the first j characters of the term v. There are also several interesting boundary conditions with regard to the Levenshtein edit distance:

<ul>
<li>The minimum value that the edit distance between two terms can take is the difference in length of the two terms</li>
<li>The maximum value of the edit distance between two terms can be the length of the term that is larger.</li>
<li>If the two terms are equal, the edit distance is zero.</li>
<li>Hamming distance between two terms is an upper bound for Levenshtein edit distance if and only if the two terms have equal length.</li>
<li>This being a distance metric also satisfies the triangle inequality property, discussed earlier when we talked about distance metrics.</li>
</ul>

There are various ways of implementing Levenshtein distance computations for terms. Here we will start with an example of two of our terms. Considering the root term 'believe' and another term 'beleive' (we ignore case in our computations). The edit distance would be 2 because we would need the following two operations:
<ul>
<li>'beleive' → 'beliive' (substitution of e to i)</li>
<li>'beliive' → 'believe' (substitution of i to e)</li>
</ul>

To implement this, we build a matrix that will basically compute the Levenshtein distance between all the characters of both terms by comparing each character of the first term with the characters of the second term. For computation, we follow a dynamic programming approach to get the edit distance between the two terms based on the last computed value. For the given two terms, the Levenshtein edit distance matrix our algorithm should generate is shown below:

<img src="https://github.com/irungus/TUDA/blob/main/img/fm8.png?raw=1" width="530" height="470">

You can see in Figure 6-1 that the edit distances are computed for each pair of characters in the terms, as mentioned earlier, and the final edit distance value highlighted in the figure gives us the actual edit distance between the two terms. This algorithm is also known as the Wagner-Fischer algorithm and is available in the paper by R. Wagner and M. Fischer titled “The String-to-String Correction Problem,” which you can refer to if you are more interested in the details.

The following function implements Levenshtein edit distance:

In [ ]:
import copy
import pandas as pd

In [ ]:
def levenshtein_edit_distance(u, v):
    # convert to lower case
    u = u.lower()
    v = v.lower()
    # base cases
    if u == v: return 0
    elif len(u) == 0: return len(v)
    elif len(v) == 0: return len(u)
    # initialize edit distance matrix
    edit_matrix = []
    # initialize two distance matrices
    du = [0] * (len(v) + 1)
    dv = [0] * (len(v) + 1)
    # du: the previous row of edit distance

    for i in range(len(du)):
        du[i] = i
    # dv : the current row of edit distances
    for i in range(len(u)):
        dv[0] = i + 1
        # compute cost as per algorithm
        for j in range(len(v)):
            cost = 0 if u[i] == v[j] else 1
            dv[j + 1] = min(dv[j] + 1, du[j + 1] + 1, du[j] + cost)
        # assign dv to du for next iteration
        for j in range(len(du)):
            du[j] = dv[j]
        # copy dv to the edit matrix
        edit_matrix.append(copy.copy(dv))
    # compute the final edit distance and edit matrix
    distance = dv[len(v)]
    edit_matrix = np.array(edit_matrix)
    edit_matrix = edit_matrix.T
    edit_matrix = edit_matrix[1:,]
    edit_matrix = pd.DataFrame(data=edit_matrix, index=list(v), columns=list(u))
    return distance, edit_matrix

That function returns both the final Levenshtein edit distance and the complete edit matrix between the two terms u and v, which are taken as input. Remember, we need to pass the terms directly in their raw string format and not their vector representations. Also, we do not consider case of strings here and convert them to lowercase.

The following snippet computes the Levenshtein edit distance between our example terms using the preceding function:

In [ ]:
for term in terms:
    edit_d, edit_m = levenshtein_edit_distance(root_term, term)
    print('Computing distance between root: {} and term: {}'.format(root_term,term))
    print('Levenshtein edit distance is {}'.format(edit_d))
    print('The complete edit distance matrix is depicted below')
    print(edit_m)
    print('-'*30)

Computing distance between root: Believe and term: beleive
Levenshtein edit distance is 2
The complete edit distance matrix is depicted below
   b  e  l  i  e  v  e
b  0  1  2  3  4  5  6
e  1  0  1  2  3  4  5
l  2  1  0  1  2  3  4
e  3  2  1  1  1  2  3
i  4  3  2  1  2  2  3
v  5  4  3  2  2  2  3
e  6  5  4  3  2  3  2
------------------------------
Computing distance between root: Believe and term: bargain
Levenshtein edit distance is 6
The complete edit distance matrix is depicted below
   b  e  l  i  e  v  e
b  0  1  2  3  4  5  6
a  1  1  2  3  4  5  6
r  2  2  2  3  4  5  6
g  3  3  3  3  4  5  6
a  4  4  4  4  4  5  6
i  5  5  5  4  5  5  6
n  6  6  6  5  5  6  6
------------------------------
Computing distance between root: Believe and term: Elephan
Levenshtein edit distance is 6
The complete edit distance matrix is depicted below
   b  e  l  i  e  v  e
e  1  1  2  3  4  5  6
l  2  2  1  2  3  4  5
e  3  2  2  2  2  3  4
p  4  3  3  3  3  3  4
h  5  4  4  4  4  4  4
a  6  

You can see from the preceding outputs that 'Believe' and 'beleive' are the closest to each other, with an edit distance of 2 and the distances between 'Believe', 'bargain', and 'Elephant' are 6, indicating a total of 6 edit operations needed. The edit distance matrices provide a more detailed insight into how the algorithm computes the distances per iteration

#### Cosine Distance and Similarity

The <code>Cosine distance</code> is a metric that can be actually derived from the Cosine similarity and vice versa. Considering we have two terms such that they are represented in their vectorized forms, <code>Cosine similarity gives us the measure of the cosine of the angle between them when they are represented as non-zero positive vectors in an inner product space</code>

Thus term vectors having similar orientation will have scores closer to 1 (cos0$^\circ$) indicating the vectors are very close to each other in the same direction (near to zero degree angle between them). Term vectors having a similarity score close to 0 (cos90$^\circ$) indicate unrelated terms with a near orthogonal angle between then. Term vectors with a similarity score close to –1 (cos180$^\circ$) indicate terms that are completely oppositely oriented to each other

<img src="https://github.com/irungus/TUDA/blob/main/img/fm9.png?raw=1" width="730" height="670">

<img src="https://github.com/irungus/TUDA/blob/main/img/fm10.png?raw=1" width="430" height="370">

Thus you can see from the position of the vectors, the plots show more clearly how the vectors are close or far apart from each other, and the cosine of the angle between them gives us the Cosine similarity metric.

In our case, we will be using the Bag of Characters vectorization to build these term vectors, and n will be the number of unique characters across the terms under analysis. An important thing to note here is that the Cosine similarity score usually ranges from –1 to +1, but if we use the Bag of Characters–based character frequencies for terms or Bag of Words–based word frequencies for documents, the score will range from 0 to 1 because the frequency vectors can never be negative, and hence the angle between the two vectors cannot exceed 90$^\circ$ .

The Cosine distance is complimentary to the similarity score can be computed by the formula,

<img src="https://github.com/irungus/TUDA/blob/main/img/fm11.png?raw=1" width="530" height="470">

where cd(u, v) denotes the Cosine distance between the term vectors u and v. The following function implements computation of Cosine distance based on the preceding formulae:

In [ ]:
def cosine_distance(u, v):
    distance = 1.0 - (np.dot(u, v) / (np.sqrt(sum(np.square(u))) * np.sqrt(sum(np.square(v)))))
    return distance

We will now test the similarity between our example terms using their Bag of Character representations, which we created earlier, available in the boc_root_vector and the boc_vector_terms variables, as depicted in the following code snippet:

In [ ]:
for term, boc_term in zip(terms, boc_vector_terms):
    print('Analyzing similarity between root: {} and term: {}'.format(root_term,term))
    distance = round(cosine_distance(root_boc_vector, boc_term),2)
    similarity = 1 - distance
    print('Cosine distance is {}'.format(distance))
    print('Cosine similarity is {}'.format(similarity))
    print('-'*40)

Analyzing similarity between root: Believe and term: beleive
Cosine distance is -0.0
Cosine similarity is 1.0
----------------------------------------
Analyzing similarity between root: Believe and term: bargain
Cosine distance is 0.82
Cosine similarity is 0.18000000000000005
----------------------------------------
Analyzing similarity between root: Believe and term: Elephan
Cosine distance is 0.35
Cosine similarity is 0.65
----------------------------------------


These vector representations do not take order of characters into account, hence the similarity between the terms "Believe" and "beleive" is 1.0 or a perfect 100 percent because it contains the same characters with the same frequency. You can see how this can be used in combination with a semantic dictionary like WordNet to provide correct spelling suggestions by suggesting semantically and syntactically correct words from a vocabulary when users type a misspelled word, by measuring the similarity between the words. You can even try our different features here instead of single character frequencies, like taking two characters at a time and computing their frequencies to build the term vectors. This takes into account some of the sequences that characters maintain in various terms. Try out different possibilities and compare the results! This distance measure works very well when measuring similarity between large documents or sentences, and we will see that in the next section when we discuss document similarity.

## Analyzing Document Similarity

We analyzed similarity between terms using various similarity and distance metrics in the previous sections. We also saw how vectorization was useful so that mathematical computations become much easier, especially when computing distances between vectors. In this section, we will try to analyze similarities between documents. By now, you must already know that a document is defined as a body of text which can be comprised of sentences or paragraphs of text. For analyzing document similarity, we will be using our utils module to extract features from document using the build_feature_ matrix() function. We will vectorize documents using their TF-IDFs similarly to what we did previously when we classified text documents or summarized entire documents. Once we have the vector representations of the various documents, we will compute similarity between the documents using several distance or similarity metrics

The
metrics we will cover in this section are as follows:
<ul>
<li>Cosine similarity</li>
<li>Hellinger-Bhattacharya distance</li>
<li>Okapi BM25 ranking</li>
</ul>



We will also test our metrics on a toy corpus here with nine documents and a separate corpus with three documents, which will be our query documents. For each of these three documents, we will try to find out the most similar documents from the corpus of nine documents, which will act as our index. Consider this to be a mini-simulation of what happens in a search
engine when you search with a sentence and the most relevant results are returned to you from its index of web pages. In our case, the queries are in the form of three documents, and relevant documents for each of these three will be returned from the index of nine documents based on similarity metrics.

In [ ]:
import numpy as np
# load the toy corpus index
toy_corpus = [
    'The sky is blue',
    'The sky is blue and beautiful',
    'Look at the bright blue sky!',
    'Python is a great Programming language',
    'Python and Java are popular Programming languages',
    'Among Programming languages, both Python and Java are the most used in Analytics',
    'The fox is quicker than the lazy dog',
    'The dog is smarter than the fox',
    'The dog, fox and cat are good friends'
    ]

In [ ]:
# load the docs for which we will be measuring similarities
query_docs = [
    'The fox is definitely smarter than the dog',
    'Java is a static typed programming language unlike Python',
    'I love to relax under the beautiful blue sky!'
    ]

From that snippet you can see that we have various documents in our corpus index that talk about the sky, programming languages, and animals. We also have three query documents for which we want to get the most relevant documents from the toy_corpus index, based on similarity computations.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

def build_feature_matrix(documents, feature_type='frequency', ngram_range=(1, 1), min_df=0.0, max_df=1.0):
    feature_type = feature_type.lower().strip()

    if feature_type == 'binary':
        vectorizer = CountVectorizer(binary=True, min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    elif feature_type == 'frequency':
        vectorizer = CountVectorizer(binary=False, min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    elif feature_type == 'tfidf':
        vectorizer = TfidfVectorizer(min_df=min_df, max_df=max_df, ngram_range=ngram_range)

    else:
        raise Exception("Wrong feature type entered. Possible values:'binary', 'frequency','tfidf'")

    feature_matrix = vectorizer.fit_transform(documents).astype(float)

    return vectorizer, feature_matrix

In [ ]:
import warnings
warnings.filterwarnings("ignore")
from nltk.corpus import stopwords
import nltk
import re

from nltk.tokenize import RegexpTokenizer
from nltk.stem import WordNetLemmatizer,PorterStemmer
lemmatizer = WordNetLemmatizer()
stemmer = PorterStemmer()
from pattern.en import suggest
from nltk.corpus import wordnet

#loading_the_stop_words_from_nltk_library_
stop_words = set(stopwords.words('english'))#e.g. an apple => "an" is a stop word. a book => "a" is a stop word

def text_preprocessing(total_text):
    if type(total_text) is not int:
        string = ""

        #replace every special char with space
        total_text = re.sub('[^a-zA-Z0-9\n]', ' ', total_text)

        #replace multiple spaces with single space
        total_text = re.sub('\s+',' ', total_text)

        #converting all the chars into lower case
        total_text = total_text.lower()

        #word tokenization
        tokenizer = RegexpTokenizer(r'\w+')
        tokens = tokenizer.tokenize(total_text)

        #Stemming
        stem_words=[stemmer.stem(w) for w in tokens]

        #Lemmatization
        lemma_words=[lemmatizer.lemmatize(w) for w in stem_words]

        for x in range(len(lemma_words)):
            #if_the_word_is_a_not_a_stop_word_then_retain_that_word_from_the_data
            if not lemma_words[x] in stop_words:

                #Repeating Characters
                #lemma_words = remove_repeated_characters(lemma_words[x])

                #Correcting incorrect or wrong spellings
                #lemma_words = suggest(lemma_words)

                string += lemma_words[x] + " "

        return string

ModuleNotFoundError: No module named 'pattern.en'

In [ ]:
# normalize and extract features from the toy corpus
norm_corpus = []
for row in range(len(toy_corpus)):
    norm_corpus.append(text_preprocessing(toy_corpus[row]))

In [ ]:
print(norm_corpus)

In [ ]:
tfidf_vectorizer, tfidf_features = build_feature_matrix(norm_corpus, feature_type='tfidf', ngram_range=(1, 1), min_df=0.0,
                                                        max_df=1.0)

In [ ]:
print(tfidf_vectorizer)

In [ ]:
print(tfidf_features)

In [ ]:
# normalize and extract features from the query corpus
norm_query_docs = []
for row in range(len(query_docs)):
    norm_query_docs.append(text_preprocessing(query_docs[row]))

In [ ]:
query_docs_tfidf = tfidf_vectorizer.transform(norm_query_docs)

In [ ]:
print(query_docs_tfidf)

Now that we have our documents normalized and vectorized with TF-IDF–based vector representations, we will look at how to compute similarity for each of the metrics, namely:
<ul>
    <li>Cosine similarity</li>
    <li>Hellinger-Bhattacharya distance</li>
    <li>Okapi BM25 ranking</li>
</ul>


## Cosine Similarity

We have seen the concepts with regards to computing Cosine similarity and also implemented the same for term similarity. Here, we will reuse the same concepts to compute the Cosine similarity scores for documents instead of terms. The document vectors will be the Bag of Words model–based vectors with TF-IDF values instead of term frequencies. We have also taken only unigrams here, but you can experiment with bigrams and so on as document features during the vectorization process. For each of the three query documents, we will compute its similarity with the nine documents in toy_corpus and return the n most similar documents where n is a user input parameter.

We will define a function that will take in the vectorized corpus and the document corpus for which we want to compute similarities. We will get the similarity scores using the dot product operation as before and finally we will sort them in reverse order and get the top n documents with the highest similarity score. The following function implements this:

In [ ]:
def compute_cosine_similarity(doc_features, corpus_features, top_n=3):
    # get document vectors
    doc_features = doc_features.toarray()[0]
    corpus_features = corpus_features.toarray()
    # compute similarities
    similarity = np.dot(doc_features, corpus_features.T)
    # get docs with highest similarity scores
    top_docs = similarity.argsort()[::-1][:top_n]
    top_docs_with_score = [(index, round(similarity[index], 3)) for index in top_docs]
    return top_docs_with_score

In that function, corpus_features are the vectorized documents belonging to the toy_corpus index from which we want to retrieve similar documents. These documents will be retrieved on the basis of their similarity score with doc_features, which basically represents the vectorized document belonging to each of the query_docs, as shown in the following snippet:

In [ ]:
print('Document Similarity Analysis using Cosine Similarity')
print('='*60)
print('\n')

for index, doc in enumerate(query_docs):
    doc_tfidf = query_docs_tfidf[index]
    top_similar_docs = compute_cosine_similarity(doc_tfidf,tfidf_features,top_n=3)
    print('Document',index+1 ,':', doc)
    print('Top', len(top_similar_docs)+1, 'similar docs:')

    print('+++'*20)

    for doc_index, sim_score in top_similar_docs:
        print('Doc num: {} Similarity Score: {}\nDoc: {}'.format(doc_index+1,sim_score, toy_corpus[doc_index]))
        print('-'*40)

    print('\n'*2)

The preceding output depicts the top two most relevant documents for each of the query documents based on Cosine similarity scores, and you can see that the outputs are quite what were expected. Documents about animals are similar to the document
that mentions the fox and the dog; documents about Python and Java are most similar to the query document talking about them; and the beautiful blue sky is indeed similar to documents that talk about the sky being blue and beautiful!

Also note the Cosine similarity scores in the preceding outputs, where 1.0 indicates perfect similarity, 0.0 indicates no similarity, and any score between them indicates some level of similarity based on how large that score is. For instance, in the last example, the main document vectors are ['sky', 'blue', 'beautiful'] and because they all match with the first document from the toy corpus, we get a 1.0 or 100 percent similarity score, and only ['sky', 'blue'] match from the second most similar document, and we get a 0.72 or 72 percent similarity score. And you should remember our discussion from earlier where I mentioned briefly that Cosine similarity using Bag of Words–based vectors only looks at token weights and does not consider order or sequence of the terms, which is quite desirable in large documents because the same content may be depicted in different ways, and capturing sequences there might lead to loss of information due to unwanted mismatches.

We recommend using scikit-learn’s cosine_similarity() utility function, which you can find under the sklearn.metrics.pairwise module. It uses similar logic as our implementation but is much more optimized and performs well on large corpora of documents. You can also use gensim’s similarities module or the cossim() function directly available in the gensim.matutils module.

## Hellinger-Bhattacharya Distance

The Hellinger-Bhattacharya distance (HB-distance) is also called the Hellinger distance or the Bhattacharya distance. The Bhattacharya distance, originally introduced by A. Bhattacharya, is used to measure the similarity between two discrete or continuous probability distributions.

HB-distance is computable for both continuous and discrete probability distributions. In our case, we will be using the TF-IDF–based vectors as our document distributions. This makes it discrete distributions because we have specific TF-IDF values for specific feature terms, unlike continuous distributions.

 As with the previous computation of Cosine similarity, we will build our function on the same principles; basically we will accept as input a corpus of document vectors and a single document vector for which we want to get the n most similar
documents from the corpus based on their HB-distances. The function implements the preceding concepts in Python in the following snippet:

In [ ]:
def compute_hellinger_bhattacharya_distance(doc_features, corpus_features, top_n=3):
    # get document vectors
    doc_features = doc_features.toarray()[0]
    corpus_features = corpus_features.toarray()
    # compute hb distances
    distance = np.hstack(np.sqrt(0.5 *np.sum(np.square(np.sqrt(doc_features) - np.sqrt(corpus_features)), axis=1)))
    # get docs with lowest distance scores
    top_docs = distance.argsort()[:top_n]
    top_docs_with_score = [(index, round(distance[index], 3)) for index in top_docs]
    return top_docs_with_score

From the preceding implementation, you case see that we sort the documents based on their scores in ascending order, unlike Cosine similarity, where 1.0 indicates perfect similarity—since this is a distance metric between distributions, a value of 0 indicates perfect similarity, and higher values indicate some dissimilarity being present. We can now apply this function to our example corpora, compute their HB-distances, and see the results in the following snippet:

In [ ]:
# get Hellinger-Bhattacharya distance based similarities for our example documents

print('Document Similarity Analysis using Hellinger-Bhattacharya distance')
print('\n')

for index, doc in enumerate(query_docs):
    doc_tfidf = query_docs_tfidf[index]
    top_similar_docs = compute_hellinger_bhattacharya_distance(doc_tfidf,tfidf_features,top_n=3)
    print('Document',index+1 ,':', doc)
    print('Top', len(top_similar_docs), 'similar docs:')

    print('++'*20)

    for doc_index, sim_score in top_similar_docs:
        print('Doc num: {} Distance Score: {}\nDoc: {}'.format(doc_index+1,sim_score, toy_corpus[doc_index]))
        print('-'*40)

    print('\n'*2)

You can see from the preceding outputs that documents with lower HB-distance scores are more similar to the query documents, and the result documents are quite similar to what we obtained using Cosine similarity. Compare the results and try out these functions with larger corpora! I recommend using gensim’s hellinger() function, available in the gensim.matutils module (which uses the same logic as our preceding function) when building large-scale systems for analyzing similarity.

## Okapi BM25 Ranking

There are several techniques that are quite popular in information retrieval and search engines, including PageRank and Okapi BM25. The acronym BM stands for best matching. This technique is also known as BM25, but for the sake of completeness I refer to it as Okapi BM25, because originally although the concepts behind the BM25 function were merely theoretical, the City University in London built the Okapi Information Retrieval system in the 1980s–90s, which implemented this technique to retrieve documents on actual real-world data. This technique can also be called a framework or model based on probabilistic relevancy and was developed by several people in the 1970s–80s, including computer scientists S. Robertson and K. Jones. There are several functions that rank documents based on different factors, and BM25 is one of them. Its newer variant is BM25F; other variants include BM15 and BM25+.

The Okapi BM25 can be formally defined as a document ranking and retrieval function based on a Bag of Words–based model for retrieving relevant documents based on a user input query. This query can be itself a document containing a sentence or collection of sentences, or it can even be a couple of words. The Okapi BM25 is actually not just a single function but is a framework consisting of a whole collection of scoring functions combined together.

There are several steps we must go through to successfully implement and compute BM25 scores for documents:
<ol>
<li>Build a function to get inverse document frequency (IDF) values for terms in corpus.</li>
<li>Build a function for computing BM25 scores for query document and corpus documents.</li>
<li>Get Bag of Words–based features for corpus documents and query documents.</li>
<li>Compute average length of corpus documents and IDFs of the terms in the corpus documents using function from point 1.</li>
<li>Compute BM25 scores, rank relevant documents, and fetch the n most relevant documents for each query document
using the function in point 2.</li>
</ol>

We will start with implementing a function to extract and compute inverse document frequencies of all the terms in a corpus of documents by using its Bag of Words features, which will contain the term frequencies, and then convert them to IDFs using the formula mentioned earlier. The following function implements this:

In [ ]:
import scipy.sparse as sp
def compute_corpus_term_idfs(corpus_features, norm_corpus):
    dfs = np.diff(sp.csc_matrix(corpus_features, copy=True).indptr)
    dfs = 1 + dfs # to smoothen idf later
    total_docs = 1 + len(norm_corpus)
    idfs = 1.0 + np.log(float(total_docs) / dfs)
    return idfs

We will now implement the main function for computing BM25 score for all the documents in our corpus based on the query document and retrieving the top n relevant documents from the corpus based on their BM25 score. The following function implements the BM25 scoring framework:

In [ ]:
def compute_bm25_similarity(doc_features, corpus_features,corpus_doc_lengths, avg_doc_length,term_idfs, k1=1.5, b=0.75, top_n=3):
    # get corpus bag of words features
    corpus_features = corpus_features.toarray()
    # convert query document features to binary features
    # this is to keep a note of which terms exist per document
    doc_features = doc_features.toarray()[0]
    doc_features[doc_features >= 1] = 1

    # compute the document idf scores for present terms
    doc_idfs = doc_features * term_idfs
    # compute numerator expression in BM25 equation
    numerator_coeff = corpus_features * (k1 + 1)
    numerator = np.multiply(doc_idfs, numerator_coeff)
    # compute denominator expression in BM25 equation
    denominator_coeff = k1 * (1 - b +(b * (corpus_doc_lengths /avg_doc_length)))

    denominator_coeff = np.vstack(denominator_coeff)
    denominator = corpus_features + denominator_coeff
    # compute the BM25 score combining the above equations
    bm25_scores = np.sum(np.divide(numerator,denominator),axis=1)

    # get top n relevant docs with highest BM25 score
    top_docs = bm25_scores.argsort()[::-1][:top_n]
    top_docs_with_score = [(index, round(bm25_scores[index], 3))for index in top_docs]
    return top_docs_with_score

In [ ]:
# build bag of words based features first
vectorizer, corpus_features = build_feature_matrix(norm_corpus,feature_type='frequency')
query_docs_features = vectorizer.transform(norm_query_docs)
# get average document length of the corpus (avgdl)
doc_lengths = [len(doc.split()) for doc in norm_corpus]
avg_dl = np.average(doc_lengths)

# Get the corpus term idfs
corpus_term_idfs = compute_corpus_term_idfs(corpus_features,norm_corpus)

# analyze document similarity using BM25 framework
print('Document Similarity Analysis using BM25')
print('\n')

for index, doc in enumerate(query_docs):
    doc_features = query_docs_features[index]
    top_similar_docs = compute_bm25_similarity(doc_features,corpus_features,doc_lengths,avg_dl,corpus_term_idfs,k1=1.5,
                                               b=0.75,top_n=4)

    print('Document',index+1 ,':', doc)
    print('Top', len(top_similar_docs), 'similar docs:')
    print('+'*40)

    for doc_index, sim_score in top_similar_docs:
        print('Doc num: {} BM25 Score: {}\nDoc: {}'.format(doc_index+1,sim_score, toy_corpus[doc_index]))
        print('-'*40)

    print('\n'*2)

You can now see how for each query document, we get expected and relevant documents that have similar concepts just like the query documents. You can see that the results are quite similar to the previous methods—because, of course, they are all similarity and ranking metrics and are expected to return similar results. Notice the BM25 scores of the relevant documents. The higher the score, the more relevant is the document. Unfortunately, I was not able to find any production-ready scalable implementation of the BM25 ranking framework in nltk or scikit-learn. However, gensim seems to have a bm25 module under the gensim.summarization package and if you are interested you can give it a try. But the core of the algorithm is based on what we implemented, and this should work pretty well on its own!